src/util/pointer.h

```diff
@@ -22,126 +22,204 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include "irrlichttypes.h"
 #include "debug.h" // For assert()
 #include <cstring>
-#include <memory> // std::shared_ptr
+#include <iterator>
+#include <memory>
+#include <sstream>
+#include <type_traits>
 
 
-template<typename T> class ConstSharedPtr {
-public:
-	ConstSharedPtr(T *ptr) : ptr(ptr) {}
-	ConstSharedPtr(const std::shared_ptr<T> &ptr) : ptr(ptr) {}
+template <typename T>
+struct is_unbounded_array : std::false_type {};
+template <typename T>
+struct is_unbounded_array<T[]> : std::true_type {};
 
-	const T* get() const noexcept { return ptr.get(); }
-	const T& operator*() const noexcept { return *ptr.get(); }
-	const T* operator->() const noexcept { return ptr.get(); }
+// Whether a `buffer<From>` can be implicitly casted to `buffer<To>`, where `buffer`
+// is some buffer smart-ptr.
+// We only allow to implicitly cast from non-const to const. If you need other
+// conversions, cast the underlying ptrs.
+template <typename From, typename To>
+using is_implicitly_array_convertible = std::integral_constant<bool,
+		std::is_same<From, To>::value
+		|| std::is_same<From, std::remove_const_t<To>>::value
+	>;
 
-private:
-	std::shared_ptr<T> ptr;
-};
+// Use this for default-initialization (aka don't fill with zeros)
+template <typename T, std::enable_if_t<!std::is_array<T>::value, bool> = true>
+std::unique_ptr<T> make_unique_for_overwrite()
+{
+	return std::unique_ptr<T>(new T);
+}
```

Comment:
What's the benefit of a separate function rather than using it as a constructor for UniqueBuffer?

Reply:
This is just `std::make_unique_for_overwrite` from C++17.

I guess you wanted to comment on `make_buffer` or `make_buffer_for_overwrite`.
One benefit is the similarity to the std lib (=> more uniformity). Also, you can decide to default- or value-init.